### PR TITLE
add topk typed implementation

### PR DESCRIPTION
**Pull request by o1lo01ol1o**
Assuming doctests don't balk at the test, this should be functional.   I needed a `FromJust` type family to work with the existing families which is maybe more principled than writing a dedicated type family at the cost of exposing unnecessary logic.

### CODE REVIEW

File: hasktorch/src/Torch/Typed/Functional.hs [hunk 1]
```diff
@@ -3905,8 +3905,24 @@ median' input = unsafePerformIO $ ATen.cast3 ATen.Managed.median_tlb
...
+-- (Tensor Float [2,3] [[ 1.0000   ,  1.0000   ,  1.0000   ],
+--                     [ 1.0000   ,  1.0000   ,  1.0000   ]],Tensor Int64 [2,3] [[ 0,  1,  2],
+--                     [ 0,  1,  2]])
+topk 
+  :: forall k dim shape dtype device 
+   . (KnownNat k, KnownNat dim, All KnownNat shape, DimOutOfBoundCheck shape dim, FromJust (ExtractDim dim shape) >= k) 
+   => Tensor device dtype shape 
+   -> Bool -- ^ if we're returning the top k largest (or top k smallest)
+   -> Bool -- ^ if the resulting k elements are themselves sorted
+   -> (Tensor device dtype (FromJust (ReplaceDim dim shape k)), Tensor device dtype (FromJust (ReplaceDim dim shape 1)))
```
Comment (tscholak) [inline hasktorch/src/Torch/Typed/Functional.hs]: nice!
alternatively, you could have defined a `TopKCheck` type family, maybe similar to `ChunkCheck`,
```Haskell
type family TopKCheck (shape :: [Nat]) (dim :: Nat) (result :: Maybe a) :: a where
  TopKCheck shape dim Nothing       = DimOutOfBound shape dim
  TopKCheck _     _   (Just result) = result
```
with
```Haskell
type TopK k shape dim = TopKCheck shape dim (ReplaceDim dim shape k)
```
This allows for customization of the error message the compiler produces, to the benefit of the user.

Comment (o1lo01ol1o) [reply hasktorch/src/Torch/Typed/Functional.hs]: Ah, yes, that's helpful, I wasn't sure what the threshold was for introducing new type-level ops.  I'll add that in `Functional.hs`

File: hasktorch/src/Torch/Typed/Functional.hs [hunk 2]
```diff
@@ -3905,8 +3905,32 @@ median' input = unsafePerformIO $ ATen.cast3 ATen.Managed.median_tlb
 -- argsort :: Tensor device dtype shape -> Int -> Bool -> Tensor device dtype shape
 -- argsort _input _dim _descending = unsafePerformIO $ (ATen.cast3 ATen.Managed.argsort_tlb) _input _dim _descending
 
--- topk :: Tensor device dtype shape -> Int -> Int -> Bool -> Bool -> (Tensor device dtype shape,Tensor device dtype shape)
--- topk _input _k _dim _largest _sorted = unsafePerformIO $ (ATen.cast5 ATen.Managed.topk_tllbb) _input _k _dim _largest _sorted
+type family TopKCheck (k :: Nat) (shape :: [Nat]) (dim :: Nat) (satd :: Maybe Nat) (result :: Maybe a) :: a where
+  TopKCheck _ shape dim _ Nothing       = DimOutOfBound shape dim
+  TopKCheck _ shape dim Nothing _       = DimOutOfBound shape dim
+  TopKCheck k shape dim (Just v) (Just result) = If ( k <=? v ) result (TypeError (Text "k must be less than or equal to the number of elements in the requested dimention and greater than or equal to 1"))
```
Comment (tscholak) [inline hasktorch/src/Torch/Typed/Functional.hs]: awesome, this is even better! though, have you tried k=0? it should just work and return an empty tensor.
there’s a typo in the error message, “dimension”.

Comment (o1lo01ol1o) [reply hasktorch/src/Torch/Typed/Functional.hs]: fixed typo and now supports k = 0

**Comment by tscholak** [comment]
Thank you @o1lo01ol1o!
I it occurred to me that there is one more thing to clarify: dtype support.
It’s likely that topk won’t work for half precision tensors on cpu, but will work on cuda.
It’s also likely that topk is not implemented for Boolean tensors.
Can you check please?

**Comment by o1lo01ol1o** [comment]
> Thank you @o1lo01ol1o!
> I it occurred to me that there is one more thing to clarify: dtype support.
> It’s likely that topk won’t work for half precision tensors on cpu, but will work on cuda.
> It’s also likely that topk is not implemented for Boolean tensors.
> Can you check please?

Ah right.  `Half` and `Bool` `CPUTensors` fail.  I don't presently have access to a cuda machine and I'm not sure where to look in the `Aten` source to determine support, do you?

**Comment by tscholak** [comment]
@o1lo01ol1o I have a nixos box with cuda at home. I can check later when I get back there.
I don't know if there's a simple way to check support in the ATen sources. My approach has been to run all permutations and then write a type family that encodes the support.
My suggestion would be to use `StandardFloatingPointDTypeValidation device dtype` for now. that is conservative, it allows only `Float` and `Double` on CPU and additionally `Half` on CUDA.

**Comment by o1lo01ol1o** [comment]
>  I can check later when I get back there.

Cool, when you can confirm, I'll add any additional restrictions on top of 03939dc

**Comment by o1lo01ol1o** [comment]
Though if CI has CUDA maybe there could be tests added somewhere?

**Comment by tscholak** [comment]
CI doesn't have CUDA, but there's a test suite that you can contribute to: 
https://github.com/hasktorch/hasktorch/blob/master/hasktorch/test/Torch/Typed/FunctionalSpec.hs#L637-L886
As you can see, tests are also run on CUDA if it is available.

**Comment by tscholak** [comment]
@o1lo01ol1o I've just tried `topk` for all dtypes on both cpu and cuda, and all combinations that are allowed given the `TopKDeviceAndDTypeCheck` type class work as expected.

I'd still like to add unit tests for `topk` before this gets merged. Also, it would be better I think if the order of arguments was changed such that the input tensor is in the last position. that will make function chaining easier on the eyes.

**Comment by junjihashimoto** [comment]
@o1lo01ol1o 
Thank you for your contribution!
I've marked topk.
https://github.com/hasktorch/hasktorch/issues/233

**Comment by o1lo01ol1o** [comment]
@tscholak I've added a basic test and fixed the argument order.  Let me know if anything else is needed here.

**Pull request merged**